### ISSUE TITLE
[PyTorch] Dense embedding doesn't work with double backward

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#6469]
Sparse version works, but dense doesn't.

code:
```python
class Test(torch.nn.Module):
    
    def __init__(self):
        super().__init__()
        self.embd = torch.nn.Embedding(1000, 100)
        self.dense = torch.nn.Linear(100, 1)
    
    def forward(self, inp):
        inp = self.embd(inp)
        return self.dense(inp)

test = Test()
test.cuda()
inp = Variable(torch.ones(10).long().cuda())
out = test(inp)
raw_loss = out.mean(dim=0)

loss_grad = torch.autograd.grad(outputs=raw_loss,
                         inputs=list(test.parameters()),
                         retain_graph=True, create_graph=True, only_inputs=True)
norm = sum([param.norm()**2 for param in loss_grad])
loss = raw_loss + norm

loss.backward(retain_graph=True)
```

https://discuss.pytorch.org/t/nn-embedding-doesnt-seem-to-support-double-backward/16165

**Comment by backpropper** [comment]
Anyone working on this?

**Comment by SsnL** [comment]
@backpropper We are busy with release, so there is no bandwidth from the team at the moment. But if you or someone submit a PR, I'd be happy to take a look after the release.

**Comment by backpropper** [comment]
I see. So does this mean this issue would not be fixed in the next release?

**Comment by SsnL** [comment]
Yes. This isn't really very high priority as most embedding uses I've seen are with sparse gradients.

**Comment by kshitij12345** [comment]
@SsnL Can I please give it a try ?

**Comment by soumith** [comment]
@kshitij12345 go for it

**Comment by backpropper** [comment]
But the PyTorch page says that the sparse library itself is experimental. Besides you cannot take strides in sparse tensors.

**Comment by KshiteejK** [comment]
From what I could understand, the implementation for backward have been moved to Aten in the master branch while for 0.3.1 ,it is in torch/nn/_functions/thnn/sparse.py

Also the function `embedding_dense_backward` seems to be declared and not defined.

And as for sparse as backpropper mentioned , Sparse do not have stride

Can anyone please advice and guide. I am a little bit lost here.

**Comment by SsnL** [comment]
@KshiteejK are your looking for guidance to implement this feature?

**Comment by kshitij12345** [comment]
Yes that would be really great and most appreciated. And sorry that previous message was from my other account.

**Comment by t-vi** [comment]
@kshitij12345 so here are some ideas how I might approach developping (as in take it with a grain of salt and only do what you think is useful to you. Also I would appreciate if you or the others pointed out if this is not helping or so.)

- The derivatives are governed by `torch/autograd` (rather than ATen itelf). Most of the magic happens in `tools/autograd`.
  Let's look into that directory: In `derivatives.yaml` are the derivatives for functions that aren't themselves built from differentiable bits. Embedding is in  [derivatives.yaml](https://github.com/pytorch/pytorch/blob/24d05662ead139883ac0d9545accc862196c3c75/tools/autograd/derivatives.yaml#L697)
  so we are looking for embedding_backward and we see that it `embedding` can be differentiated w.r.t. weight (only).
- Many of the backwards are defined in `tools/autograd/templates/Functions.cpp`. However, this is not the case here. So we look in the `aten` directory (I personally like to use `rgrep <function_name> aten/src/` a lot). Indeed, it is defined in [`aten/src/ATen/native/native_functions.yaml`](https://github.com/pytorch/pytorch/blob/24d05662ead139883ac0d9545accc862196c3c75/aten/src/ATen/native/native_functions.yaml#L291-L298) and the definitions are in the `Embedding.cpp` and `cuda/Embedding.cu` in that directory.
- Get a handle on what you expect to see and what the current state is.
  My repro would look something like this:
```python
input = torch.tensor([[1,2,4,5],[4,3,2,9]])
embedding_matrix = torch.rand(10, 3, requires_grad=True)
emb = torch.nn.functional.embedding(input, embedding_matrix) # optionally have this weight in backward
#emb = embedding_matrix[input] # alternative for unweighted backward
weights = torch.randn_like(emb)
l = (emb**2*weights).sum()
d_embedding_matrix, d_emb, = torch.autograd.grad(l, [embedding_matrix,emb], create_graph=True, retain_graph=True)
d_embedding_matrix.sum().backward()
```
  I usually define python functions and checking they do the same as the existing or desired derivative.

    - The forward of embedding is basically torch.index_select with a bit of `view` for non-flat indices.
    - The backward is a fusion of `_index_add` and (indeed, look at index_select in `derivatives.yaml`) an optional multiplication by the inverse frequencies. (Indeed, one might be tempted benchmark in python much slower `index_add_` and `mul_` are - if that is the case, defining a partial derivative for `mul_` and using that would solve the problem. Possibly double checking whether embedding's OPENMP ifdef might be incorporated into index_add_.) 
    - If you were to define the double backward, it would be the backward of index_add_ (which is index_select_ again) and mul (which is the frequency weighting on the grad). You can check this by using defining a derivative with `index_add_`and `mul_`. Note that for double backwards you also need the derivative by `grad` (the gradient w.r.t. the outputs).

Then you need an implementation strategy. This might also be something where the core devs and the people implementing the current code have an opinion. (@apaske, @colesbury ?)
- If we leave the backward as is, one could just implement embedding_double_backward in `tools/autograd/templates/Functions.cpp` that implements the necessary double backwards. However it might be nicer to avoid this overly specific code, and if you start before getting the input of the people that review your patch, you might end up coding for experience only. (Guess how I know...)
- Introduce an optional weight to `index_add_` and define embedding_backward with that. This would be my favourite  as it would seem the cleanest way.
- Possibly see whether `index_add_` and `mul_` separately is an option.

I hope this helps to get you started. You can also ping me directly if I can be of help.

**Comment by elanmart** [comment]
@SsnL how do you think this could be best approached? Custom double-backward? Adding something like `index_add_weighted_` as @t-vi suggested?

**Comment by SsnL** [comment]
@t-vi Thanks for the detailed explanation.

@elanmart I agree with @t-vi 's suggestion on adding optional weight matrix on `index_add_`. This would also require we output the inverse of `counts` in the `embedding_backward` functions as floating point tensors, but I think this is the cleanest solution.

**Comment by dqian96** [comment]
hey @kshitij12345 are you still working on this? can I give this a go if you don't mind? Thanks!

**Comment by kshitij12345** [comment]
@dqian96 I am still working on it. Give me 2 weeks till 8th of July and I will pass it to you if still unable to do.Till then could you also please help me 😄

**Comment by ustctf** [comment]
Hi @kshitij12345 , really looking forward and appreciated for your implementation! 

The double backward operator is becoming more and more important recently such as the usage in recent NAS work [https://arxiv.org/abs/1806.09055](url). To support double backward for Embedding is significantly desirable for NLP related tasks that are built upon these latest research.

**Comment by gchanan** [comment]
@kshitij12345 are you still working on this?

**Comment by kshitij12345** [comment]
@gchanan yes i am still on it.
@ustctf Thank you for your kind words.. I ll give my best :)

**Comment by kshitij12345** [comment]
The version on my [branch](https://github.com/kshitij12345/pytorch) solves the issue. You can follow the PR [here](https://github.com/pytorch/pytorch/pull/9078). Also if you need the functionality now, it would be great if you could try the code on my branch and let me know if you face any issues.
Thank You.
